warlords.6
.TH WARLORDS 6 2013-12-02
.SH NAME
warlords-server \- The game server.
.P
warlords-client \- The game client.
.SH SYNOPSIS
.B warlords-server
[-t timeout] [-m min_qt] [-l lobby_timeout]
.P
.B warlords-client
[-s server] [-p port] [-n name] [-m]
.SH DESCRIPTION
.B Server Options:
.IP -t
Timeout before clients have taken too long to play. The default is 15 seconds.
If more than 3 minutes are requested, the default will be used.
.IP -m
Minimum quantity of players before a game may begin.
Default is 3, this is the minimum minimum.
No more than 7 players can play;
If more than 7 is requested as the minimum, then the default will be used.
Range: [3,5].
.IP -l
Timeout before a game begins once the minimum has been reached.
The default is 15 seconds.
If more than 3 minutes are requested, the default will be used.
.IP -g
Use curses gui. Displays more information for an active server
administrator to interact with the game in progress.
.P
.B Client Options:
.IP -s
IP or DNS name for the server to connect to. Default is "localhost".
.IP -p
The port to use for connecting to the server. Default is 36737.
If a number less than 36700 is requested then 36700 will be added to it.
This way you can connect to another server in the WWU CSCI367 class using only the student's book number.
.IP -n
The name to use for your player. The default is my last name: "Vergaray".
Requests for names less than the default length will be ignored.
You must padd the end with spaces in this case. If the name is too long it will be truncated.
.IP -m
Manual mode. Allows human interaction mode. Auto-play is the default.
.SH Client Manual Mode
player must click "play" button to join.
You can also send, messages to the server using the correct protocol in the chat bar.
For example: type "[chand]" as a chat message at any time to request the server to send you your hand.
WARNING: You are responsible for propper timing and formatting of your direct messages.
Without the beginning '[', the client will send whatever you type as a chat message.
.SH Server GUI Mode
You are shown an interface with 3 major sections.
The top section displays information about the players at the table:
.P
NAME [ix] S: 3 4 5 6 7 8 9 0 J Q K A 2 (TT) Score Strikes [Hands: ]
.P
Name, index (player id), Status, the quantity of cards in each rank
and the total number of cards held by a player.
As well as, a score calculated by giving warlords +1 and scumbags -1 at the
beginning of each hand. This gets incremented at the same time as they are asked
to swap cards.
The number of strikes a player at the table has, is shown, as well as a list of the
strike codes they recieved. Finally the number of hands the player has taken part in
is shown last. In the top right corner, the total quantity of hands completed by the
server is tracked.
.P
The middle area--in a box--is the lobby.
For now it's just a list of names separated by commas.
.P
The largest section is given for watching chat messages and reciving messages from
the server as you type commands.
.P
To use
.B Server Commands
begin with either a forward or a backward slash.
Commands which affect individual players always require the player's index (the player id) as the first parameter.
See Known Bugs below for usage idiosyncrasies, to avoid experienceing unexpected behavior.
.IP /b i
Slash-b is the command for banning a player from chat. A notification will be
sent to all players, and the banned player will recieve strikes for further chat attempts.
.IP /c msg
Slash-c is for broadcasting chat messages from the server administrator.
.IP /f i
Slah-f is to forgive a player one strike.
.IP /k i
Slash-k will kick a client from being connected to the server.
.IP /s i xx
Slash-s is the command for striking a player. You can give any player a strike
for any reason any time.
It requires two parameters, first, the index of the player (the player id), as ususal.
Followed by the two-digit strike code to send to them.
.IP /u i
Slash-u is the counterpart of Slash-b. It will un-ban the player from chat.
.IP /w i msg
Slash-w is for whisper. This sends a chat message only to the specified client.
TODO: Allow clients this ability. & Show them in SGUI. Colorize them in both UIs.
.IP /X
Slash-X will shutdown the server. CTRL-C works as well.
.SH NOTES
This game was created for the CSCI367 Networking Class at WWU.
.P
A matter of some debate in class was the inclusion of the trump cards in table messages.
This server sends the table showing the trump play for the client's GUI.
The automated client doesn't care, but a human player might get confused
by another server which never shows a two being played.
.SH KNOWN BUGS
Not everything planned has been implemented:
gtk menu for the client,
advanced ai,
different warlord/scumbag images with nametags.
Grep TODO for more.
.P
.B Temporary Lobby Overpopulation
Connect 3 clients. Kill 2 of them simultaneously. Remaining player is enqueued to lobby twice.
Join another: Server thinks there's enough players and begins a game.
Then, it realizes the duplicate and re-enqueues players.
.P
.B Timeout Confusion
Lobby timeout duration expected between hands, but error timeout duration is behavior.
.P
.B Full Lobby Garbage Entry
In the client, when the lobby is full--possibly before--garbage is appended in place of the last players name.
This is probably due to reaching beyond the valid portion of the array.
.P
.B SGUI Command Line
Typing a command during play of a hand will halt play until finished.
If typing a command between hands, when a hand starts,
partially input commands will be interupted and forced to attempt to execute.
For now this is silently ignored, and the command must be retyped.
FIX IDEA: halt SIGALARM for duration of stdin input. ?
Or, use raw input with arrowkeys and highlighting to use command sequences rather than command line.


warlords.c
/* $Id: warlords.c,v 1.45 2013/12/05 04:01:58 moonsdad Exp $ */


/******************************************************************************
 * Warlords And Scumbags                                                      *
 *      warlords.c    --  The Client main file.                               *
 ******************************************************************************/
#define MAIN
#include "client.h"

char localhost[] = "localhost"; /* default host name */

/********************************************************************** MAIN: */
int main( int argc, char **argv )
{
    struct hostent  * ptrh;     /* pointer to a host table entry */
    struct protoent * ptrp;     /* pointer to a protocol table entry */
    struct sockaddr_in sad;     /* structure to hold an IP address */
    int sd;                     /* socket descriptor */
    int port = PROTOPORT;       /* protocol port number */
    char* host = localhost;     /* pointer to host name */
    IODAT io_data;              /* BUFFERs holding Game Data and IO */
    int i;

    /* GTK Variables */
    const gchar* title = "Warlords and Skumbags";
    const gchar* icon = "./dat/img/13031717912126086382istockphoto_6922629-wild-joker-in-a-deck-of-cards-th.png";
    gchar* tplist_coltitle[LIST_COL_QT] = { "STATUS", "NAME", "CARDS" };
    GIOChannel* Sd;
    GtkWidget* widget[MAX_WIDGETS]; /* WARNING: Beware the difference between *
                                     *  widget[] and Widget[]. The former has *
                                     *  global scope and a different range of *
                                     *  valid indexes.                        */

    /* Parse Command Line Options */
    opterr = 0;
    while( (i = getopt( argc, argv, "s:p:n:m" )) != -1 ) {
        switch( i ) {
            case 's':
                host = optarg;
                break;
            case 'p':
                port = atoi(optarg);
                if( port < 36700 ) port += 36700;
                break;
            case 'n':
                if( strlen( optarg ) >= MAX_PNAME_LEN )
                    strncpy( my_name, optarg, MAX_PNAME_LEN );
                my_name[MAX_PNAME_LEN] = '\0'; /* Assert null-termniated String */
                break;
            case 'm':
                auto_mode = false;
                BARF( "auto_mode disabled.", 0 );
            case '?':
                fprintf( stderr, "Ignoring " );
                switch( optopt ){
                    case('s'): case('p'): case('n'):
                        fprintf( stderr, "Option -%c : requires an argument.\n", optopt);
                        break;
                    default:
                        if( isprint (optopt) )
                            fprintf (stderr, "Unknown option `-%c'.\n", optopt);
                        else fprintf (stderr, "Unknown option character `0x%x'.\n", optopt );
                }/* End optopt Switch */
        }/* End getopt rval Switch */
    }/* End getopt While */

    /* Init IODAT */
    io_data.input.n = io_data.output.n =
    io_data.slobb.n = io_data.stabl.n  = io_data.shand.n = 0;

    swap_cd[1][0] = swap_cd[0][0] = '5';
    swap_cd[1][1] = swap_cd[0][1] = '2';
    swap_cd[1][2] = swap_cd[0][2] = '\0';


    /************************* Set up Connection ***************************/
    memset((char *)&sad,0,sizeof(sad)); /* clear sockaddr structure */
    sad.sin_family = AF_INET; /* set family to Internet */

    if( port > 0 ) /* test for legal value */
        sad.sin_port = htons((u_short)port);
    else exit( Error( "bad port number:", port ) );

    /* Convert host name to equivalent IP address and copy to sad. */
    if( (ptrh = gethostbyname( host )) )
        memcpy( &sad.sin_addr, ptrh->h_addr, ptrh->h_length );
    else exit( ERROR( host, "invalid host", 1 ));

    /* Map TCP transport protocol name to protocol number. */
    if( !(ptrp = getprotobyname("tcp")) )
        exit( Error( "Cannot map \"tcp\" to protocol number!", 1 ) );

    /* Create a socket. */
    if( (sd = socket(PF_INET, SOCK_STREAM, ptrp->p_proto)) < 0 )
        exit( ERROR( "socket", strerror(errno), sd ) );
    gsd = sd; //HACK: make sd global?

    /* Connect the socket to the specified server. */
    if( connect( sd, (struct sockaddr *)&sad, sizeof(sad) ) < 0 )
        exit( ERROR( "connect", strerror(errno), sd ) );

    /* Set socket to nonblocking */
    fcntl_setf( sd, O_NONBLOCK );

    /* Setup IO Channel for GTK Main Loop */
    if( !(Sd = g_io_channel_unix_new( sd )) )
        exit( ERROR("g_io_channel_unix_new", "failed", sd) );

    if( auto_mode ) click_pbtn(NULL, NULL);
    /*************************** Set up GUI *********************************/
    gtk_init( &argc, &argv );

    /* Add Channel to GTK Main Loop */
//     g_io_add_watch_full( Sd, G_PRIORITY_HIGH, G_IO_IN, getmsg, (gpointer)&io_data, NULL/*?_dtor*/ );
    g_io_add_watch( Sd, G_IO_IN, getmsg, (gpointer)&io_data );

    /* Setup Main Window */
    widget[MAIN_WINDOW] = gtk_window_new( GTK_WINDOW_TOPLEVEL );
    gtk_window_set_title( GTK_WINDOW (widget[MAIN_WINDOW]), title );
    gtk_window_set_icon( GTK_WINDOW (widget[MAIN_WINDOW]), load_pixbuf(icon));
    g_signal_connect( widget[MAIN_WINDOW], "destroy", G_CALLBACK (gtk_main_quit), NULL);
    gtk_container_set_border_width( GTK_CONTAINER (widget[MAIN_WINDOW]), 5 );
    gtk_window_set_default_size( GTK_WINDOW (widget[MAIN_WINDOW]), DEFAULT_WINDOW_SIZE );

    widget[MAIN_FRAME] = gtk_table_new( 6, 3, FALSE ); //gtk_vbox_new( FALSE, 0 );
    gtk_container_add( GTK_CONTAINER (widget[MAIN_WINDOW]), widget[MAIN_FRAME] );


    /* Setup Table Frame */
    widget[TABLE_FRAME] = gtk_vbox_new( FALSE, 0 );
    gtk_table_attach_defaults( GTK_TABLE (widget[MAIN_FRAME]), widget[TABLE_FRAME], 0, 2, 0, 5 );
    Widget[ADD_PLAYERS] = gtk_hbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX (widget[TABLE_FRAME]), Widget[ADD_PLAYERS], TRUE, TRUE, 0 );
    Widget[TABLE_TOP] = gtk_hbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX (widget[TABLE_FRAME]), Widget[TABLE_TOP], TRUE, TRUE, 0 );
    gtk_widget_show( Widget[TABLE_TOP] );
    gtk_widget_show( Widget[ADD_PLAYERS] );
    gtk_widget_show( widget[TABLE_FRAME] );

    /* Pack Empty Seats */
    for( i = 0; i < 4; i++ )
        gtk_box_pack_start( GTK_BOX (Widget[ADD_PLAYERS]), create_image( "./dat/img/empty_seat.png",
                NULL, NULL ), TRUE, TRUE, 0 );
    gtk_box_pack_start( GTK_BOX (Widget[TABLE_TOP]), create_image( "./dat/img/empty_seat.png",
                NULL, NULL ), TRUE, TRUE, 0 );
    for( i = 0; i < 4; i++ )
        gtk_box_pack_start( GTK_BOX (Widget[TABLE_TOP]),
            create_image( "./dat/img/cards/52.jpeg",
                NULL, NULL ), TRUE, TRUE, 0 );
    gtk_box_pack_start( GTK_BOX (Widget[TABLE_TOP]), create_image( "./dat/img/empty_seat.png",
                NULL, NULL ), TRUE, TRUE, 0 );

    /* Create Hand Window */
    widget[HAND_WINDOW] = gtk_window_new( GTK_WINDOW_TOPLEVEL );
    gtk_window_set_title( GTK_WINDOW (widget[HAND_WINDOW]), "Hand" );
    gtk_window_set_icon( GTK_WINDOW (widget[HAND_WINDOW]), load_pixbuf(icon));
    gtk_container_set_border_width( GTK_CONTAINER (widget[HAND_WINDOW]), 10 );
//     gtk_window_set_default_size( GTK_WINDOW (widget[HAND_WINDOW]), WINDOW_SIZE_HAND );
    g_signal_connect( widget[HAND_WINDOW], "destroy", G_CALLBACK (gtk_main_quit), NULL);//TODO: make recreate
    Widget[HAND] = gtk_hbox_new( FALSE, 0 );
    gtk_container_add( GTK_CONTAINER (widget[HAND_WINDOW]), Widget[HAND] );
    gtk_widget_show( Widget[HAND] );
    /* Pack Empty Hand into Hand Box */
    for( i = 0; i < MAX_PHAND_LEN; i++ )
        gtk_box_pack_start( GTK_BOX (Widget[HAND]),
            create_image( "./dat/img/cards/52.jpeg",
                G_CALLBACK(click_hcard), "52" ), TRUE, TRUE, 0 );
    gtk_widget_show( widget[HAND_WINDOW] );

    /* Create Play Window */
    widget[PLAY_WINDOW] = gtk_window_new( GTK_WINDOW_TOPLEVEL );
    gtk_window_set_title( GTK_WINDOW (widget[PLAY_WINDOW]), "Play" );
    gtk_window_set_icon( GTK_WINDOW (widget[PLAY_WINDOW]), load_pixbuf(icon));
    gtk_container_set_border_width( GTK_CONTAINER (widget[PLAY_WINDOW]), 10 );
//     gtk_window_set_default_size( GTK_WINDOW (widget[PLAY_WINDOW]), WINDOW_SIZE_PLAY );
    g_signal_connect( widget[PLAY_WINDOW], "destroy", G_CALLBACK (gtk_main_quit), NULL);//TODO: make recreate
    if( !auto_mode ) gtk_widget_show( widget[PLAY_WINDOW] );

    widget[PLAY_FRAME] =  gtk_vbox_new( FALSE, 0 );
    gtk_container_add( GTK_CONTAINER (widget[PLAY_WINDOW]), widget[PLAY_FRAME] );
    gtk_widget_show( widget[PLAY_FRAME] );
    widget[BTN_FRAME] = gtk_hbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX (widget[PLAY_FRAME]), widget[BTN_FRAME], TRUE, TRUE, 0 );
    gtk_widget_show( widget[BTN_FRAME] );

    widget[PLAY_BUTTON] = gtk_button_new_with_label("PLAY");
    g_signal_connect( widget[PLAY_BUTTON], "clicked", G_CALLBACK (click_pbtn), "pl" );
    gtk_box_pack_start( GTK_BOX (widget[BTN_FRAME]), widget[PLAY_BUTTON], FALSE, FALSE, 4);
    gtk_widget_show( widget[PLAY_BUTTON] );

    widget[PASS_BUTTON] = gtk_button_new_with_label("PASS");
    g_signal_connect( widget[PASS_BUTTON], "clicked", G_CALLBACK (click_pbtn), "ps" );
    gtk_box_pack_start( GTK_BOX (widget[BTN_FRAME]), widget[PASS_BUTTON], FALSE, FALSE, 4);
    gtk_widget_show( widget[PASS_BUTTON] );

    widget[CLR_BUTTON] = gtk_button_new_with_label("CLEAR");
    g_signal_connect( widget[CLR_BUTTON], "clicked", G_CALLBACK (click_pbtn), "cl" );
    gtk_box_pack_start( GTK_BOX (widget[BTN_FRAME]), widget[CLR_BUTTON], FALSE, FALSE, 4);
    gtk_widget_show( widget[CLR_BUTTON] );

    Widget[PLAY] = gtk_vbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX (widget[PLAY_FRAME]), Widget[PLAY], TRUE, TRUE, 0 );
    gtk_widget_show( Widget[PLAY] );
    for( i = 0; i < 4; i++ )
        gtk_box_pack_start( GTK_BOX (Widget[PLAY]), create_image( "./dat/img/cards/52.jpeg",
                NULL, NULL ), TRUE, TRUE, 0 );

    /* Setup Player List View */
    widget[STAT_VIEW] = gtk_scrolled_window_new( NULL, NULL );
    gtk_scrolled_window_set_policy( GTK_SCROLLED_WINDOW (widget[STAT_VIEW]), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
    gtk_table_attach_defaults( GTK_TABLE(widget[MAIN_FRAME]), widget[STAT_VIEW], 2, 3, 0, 6 );
    widget[STAT_FRAME] = gtk_vbox_new( FALSE, 0 );

    /* Setup Table Player List */
    playerlist[TABBLL] = gtk_list_store_new( LIST_COL_QT, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_STRING );
    widget[TABLE_TREE] = gtk_tree_view_new_with_model( GTK_TREE_MODEL (playerlist[TABBLL]) );
    for( i = 0; i < LIST_COL_QT; i++ )
        gtk_tree_view_append_column( GTK_TREE_VIEW (widget[TABLE_TREE]),
            gtk_tree_view_column_new_with_attributes( tplist_coltitle[i],
                gtk_cell_renderer_text_new(), "text", i, NULL ) );
    gtk_widget_show( widget[TABLE_TREE] );
    gtk_box_pack_start( GTK_BOX (widget[STAT_FRAME]), widget[TABLE_TREE], TRUE, TRUE, 0 );
    gtk_widget_show( widget[STAT_FRAME] );

    /* Setup Lobby Player List */
    playerlist[LOBBYL] = gtk_list_store_new( 1, G_TYPE_STRING );
    widget[LOBBY_TREE] = gtk_tree_view_new_with_model( GTK_TREE_MODEL (playerlist[LOBBYL]) );
    {   GtkTreeViewColumn* l_col = gtk_tree_view_column_new_with_attributes( "The Lobby",
            gtk_cell_renderer_text_new(), "text", 0, NULL );
        gtk_tree_view_column_set_alignment( l_col, 0.5 );
        gtk_tree_view_append_column( GTK_TREE_VIEW (widget[LOBBY_TREE]), l_col);
    } gtk_widget_show( widget[LOBBY_TREE] );
    gtk_box_pack_start( GTK_BOX (widget[STAT_FRAME]), widget[LOBBY_TREE], TRUE, TRUE, 0 );
    gtk_widget_show( widget[STAT_FRAME] );

    gtk_scrolled_window_add_with_viewport( GTK_SCROLLED_WINDOW (widget[STAT_VIEW]), widget[STAT_FRAME] );
    gtk_widget_show( widget[STAT_VIEW] );

    /* Setup Text Chat/Msg Frame */
    widget[COM_FRAME] = gtk_vbox_new( FALSE, 0 );
    gtk_table_attach_defaults( GTK_TABLE(widget[MAIN_FRAME]), widget[COM_FRAME], 0, 2, 5, 6 );
    gtk_widget_show( widget[COM_FRAME] );

    widget[MESG_BOX] = gtk_scrolled_window_new( NULL, NULL );
    gtk_scrolled_window_set_policy( GTK_SCROLLED_WINDOW (widget[MESG_BOX]), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC );
    gtk_box_pack_start( GTK_BOX (widget[COM_FRAME]), widget[MESG_BOX] , TRUE, TRUE, 4);
    gtk_widget_show( widget[MESG_BOX] );

    widget[MESG_FIELD] = gtk_text_view_new();
    display_buf = gtk_text_view_get_buffer( GTK_TEXT_VIEW (widget[MESG_FIELD]) );
    gtk_text_view_set_editable( GTK_TEXT_VIEW (widget[MESG_FIELD]), FALSE );
    gtk_text_view_set_cursor_visible( GTK_TEXT_VIEW (widget[MESG_FIELD]), FALSE );
    gtk_text_view_set_wrap_mode( GTK_TEXT_VIEW (widget[MESG_FIELD]), GTK_WRAP_WORD );
    gtk_widget_show( widget[MESG_FIELD] );
    gtk_scrolled_window_add_with_viewport( GTK_SCROLLED_WINDOW (widget[MESG_BOX]), widget[MESG_FIELD] );

    widget[CHAT_FIELD] = gtk_text_view_new();
    gtk_text_view_set_editable( GTK_TEXT_VIEW (widget[CHAT_FIELD]), TRUE );
    gtk_text_view_set_cursor_visible( GTK_TEXT_VIEW (widget[CHAT_FIELD]), TRUE );
    gtk_text_view_set_wrap_mode( GTK_TEXT_VIEW (widget[CHAT_FIELD]), GTK_WRAP_WORD );
    g_signal_connect( widget[CHAT_FIELD], "key_press_event", G_CALLBACK(chat_key_press), (gpointer)widget[CHAT_FIELD] );
    widget[CHAT_BOX] = gtk_hbox_new( FALSE, 0 );
    gtk_box_pack_start( GTK_BOX (widget[COM_FRAME]),  widget[CHAT_BOX] , FALSE, FALSE, 4);
    widget[SAY_BTN] = gtk_button_new_with_label("SAY:");
    g_signal_connect( widget[SAY_BTN], "clicked", G_CALLBACK (parse_chat), (gpointer)widget[CHAT_FIELD] );
    g_signal_connect_after( widget[SAY_BTN], "clicked", G_CALLBACK (send_mesg), (gpointer)Sd );
    gtk_box_pack_start( GTK_BOX (widget[CHAT_BOX]), widget[SAY_BTN], FALSE, FALSE, 4);
    gtk_box_pack_start( GTK_BOX (widget[CHAT_BOX]), widget[CHAT_FIELD], TRUE, TRUE, 4);
    gtk_widget_show( widget[SAY_BTN] );
    gtk_widget_show( widget[CHAT_BOX] );
    gtk_widget_show( widget[CHAT_FIELD] );

    /* Display GUI */
    gtk_widget_show( widget[MAIN_FRAME] );
    gtk_widget_show( widget[MAIN_WINDOW] );

    gtk_main();

    close(sd);
    return 0;
}/* End main Func */


/******************************************************************************/
/* Function:    click_hcard             G_CALLBACK                            */
/* Parameters:  GtkWidget*      event_box                                     */
/*              GdkEventButton* event                                         */
/*              gpointer        data                                          */
/* Returns:     gboolean                TRUE if event handling completed.     */
/*                                      FALSE to continue invoking callbacks. */
/* This is the handler for when cards are clicked. It directs the call to     */
/*      click_cswap if iam_warlord -- The card is selected for swaping.       */
/* Otherwise, The data is either a string with:                               */
/*      "ps"    -- Player is Passing.                                         */
/*      "pl"    -- Player is Playing.                                         */
/*      "cl"    -- Player is Clearing Current Prepared Play.                  */
/*      card    -- Number representing Card Face.                             */
/******************************************************************************/
gboolean click_hcard( GtkWidget *event_box, GdkEventButton *event, gpointer data )
{
    static char playmsg[20] = "[cplay|52,52,52,52]";
    static int cards_played = 0;
    char fp[24] = "./dat/img/cards/52.jpeg",
        *Data   = (char*)data; /* Get Rid of Warning */

    assert( data != NULL );

    if( iam_warlord ) return click_cswap( NULL, NULL, data );

#ifdef DEBUG_BARF
    /* BARF */
    if( event )
        g_print( "Event box clicked at coordinates %f,%f", event->x, event->y );
    g_print( "== Card # %s\n", (char*)data );
#endif

    switch( Data[0] ) {
        case 'p':
            if( Data[1] == 's' ) memcpy( playmsg+7, "52,52,52,52", 11 );
            if( Write( gsd, playmsg, 19 ) < 0 )
                ERROR( "write", strerror(errno), gsd );
            #ifdef DEBUG_BARF
                g_print( "PLAYED: %s", playmsg );//BARF
            #endif
            is_myturn = false;
            if( Write( gsd, "[chand]", 7 ) < 0 )
                ERROR( "write", strerror(errno), gsd );
            /* NO BREAK */
        case 'c' :
            cards_played = 0;
            memcpy( playmsg+7, "52,52,52,52", 11 );
            break;
        default:
            if( !isdigit( Data[0] ) ) break; //Assert no Garbage From GTK
            memcpy( playmsg+7+(3*cards_played), data, 2 );
            if( ++cards_played == 4 ) cards_played = 0; /* Loop Around */
            //TODO: have preference setting for other behaviour
    }/* End Data Switch */

    /* Display Play So Far */
    gtk_container_foreach( GTK_CONTAINER (Widget[PLAY]), widget_destroy, NULL );
    fp[16] = playmsg[7]; fp[17] = playmsg[8];
    gtk_box_pack_start( GTK_BOX (Widget[PLAY]),
        create_image( fp, NULL, NULL ), TRUE, TRUE, 0 );
    fp[16] = playmsg[10]; fp[17] = playmsg[11];
    gtk_box_pack_start( GTK_BOX (Widget[PLAY]),
        create_image( fp, NULL, NULL ), TRUE, TRUE, 0 );
    fp[16] = playmsg[13]; fp[17] = playmsg[14];
    gtk_box_pack_start( GTK_BOX (Widget[PLAY]),
        create_image( fp, NULL, NULL ), TRUE, TRUE, 0 );
    fp[16] = playmsg[16]; fp[17] = playmsg[17];
    gtk_box_pack_start( GTK_BOX (Widget[PLAY]),
        create_image( fp, NULL, NULL ), TRUE, TRUE, 0 );

    return TRUE;
}/* End click_hcard Func */


/******************************************************************************/
/* Function:    click_cswap             G_CALLBACK                            */
/* Parameters:  GtkWidget*      event_box                                     */
/*              GdkEventButton* event                                         */
/*              gpointer        data                                          */
/* Returns:     gboolean                TRUE if event handling completed.     */
/*                                      FALSE to continue invoking callbacks. */
/* This is the handler for when cards are clicked for swaping.                */
/* data is a character array beginning with an 's' for SWAP, or containing    */
/*      a card number for the card being selected for swapping.               */
/******************************************************************************/
gboolean click_cswap( GtkWidget *event_box, GdkEventButton *event, gpointer data )
{
    static char playmsg[11] = "[cswap|52]";
    char fp[24] = "./dat/img/cards/52.jpeg",
        *Data   = (char*)data; /* Get Rid of Warning */

    assert( data != NULL );

    if( !iam_warlord ) return TRUE;

    if( Data[0] == 's' ) {
        if( Write( gsd, playmsg, 10 ) < 0 )
                ERROR( "write", strerror(errno), gsd );
        if( Write( gsd, "[chand]", 7 ) < 0 )
                ERROR( "write", strerror(errno), gsd );
        iam_warlord = false;
    } else {
        memcpy( playmsg+7, data, 2 );
        gtk_container_foreach( GTK_CONTAINER (Widget[SWAP]), widget_destroy, NULL );
        fp[16] = swap_cd[0][0]; fp[17] = swap_cd[0][1];
        gtk_box_pack_start( GTK_BOX (Widget[SWAP]),
            create_image( fp, G_CALLBACK(click_cswap), swap_cd[0] ), TRUE, TRUE, 0 );
        fp[16] = Data[0]; fp[17] = Data[1];
        gtk_box_pack_start( GTK_BOX (Widget[SWAP]),
            create_image( fp, NULL, swap_cd[0] ), TRUE, TRUE, 0 );
    }/* End Else */
    return TRUE;
}/* End click_cswap Func */


/******************************************************************************/
/* Function:    click_pbtn              G_CALLBACK                            */
/* Parameters:  GtkWidget*      button                                        */
/*              gpointer        data                                          */
/* Returns:     gboolean                TRUE if event handling completed.     */
/*                                      FALSE to continue invoking callbacks. */
/* This is the handler wrapper for when the "play" button is clicked.         */
/*  If player has joined, pass data on to click_hcard. Else send a cjoin.     */
/* TODO: have popup menu for last minute configuration if non auto_mode.      */
/******************************************************************************/
gboolean click_pbtn( GtkWidget *button, gpointer data ) {
    char cjoin[16];

    if( iam_joined ) return click_hcard( NULL, NULL, data );

    sprintf( cjoin, "[cjoin|%s]", my_name );
    if( Write( gsd, cjoin, 16 ) < 0 )
        ERROR( "write", strerror(errno), gsd );
    return TRUE;
}/* End click_pbtn Func */


/******************************************************************************/
/* Function:    chat_key_press              G_CALLBACK                        */
/* Parameters:  GtkWidget*      widget                                        */
/*              GdkEventKey*    key                                           */
/*              gpointer        data                                          */
/* Returns:     gboolean                TRUE if event handling completed.     */
/*                                      FALSE to continue invoking callbacks. */
/* This is the handler for chat message box input. It detects an 'enter' key  */
/* press as the signal to parse the chat message and send it on over.         */
/* I'm unsure of the gtk constant which matches the hex keyval.               */
/******************************************************************************/
gboolean chat_key_press( GtkWidget* widget, GdkEventKey* key, gpointer data )
{
    if(( key->type == GDK_KEY_PRESS )&&( key->keyval == 0xff0d )) {//HACK
        parse_chat( NULL, data );
        if( Write( gsd, transmit.buf, transmit.n ) < 0 )
                    ERROR( "write", strerror(errno), -1 );
        transmit.n = 0;
        return true;
    } else return false;
}/* End chat_key_press Func */


/************************************EOF***************************************/
